File: src/Api/Helpers/FileHelper.cs
/*
* Copyright (C) Rodrigo Ferreira, All Rights Reserved
* Unauthorized copying of this file, via any medium is strictly prohibited
* Proprietary and confidential
*/

using System.Text.RegularExpressions;

namespace FaceDetector.Api.Helper;

public class FileHelper
{
    public static string GetContentType(IFormFile file)
    {
        // Try to get the content type from the file itself
        if (!string.IsNullOrEmpty(file.ContentType))
        {
            return file.ContentType.ToLower();
        }

        // If content type is not available, try to determine it from the extension
        string extension = Path.GetExtension(file.FileName).ToLowerInvariant();

        return extension switch
        {
            ".jpg" or ".jpeg" => "image/jpeg",
            ".png" => "image/png",
            ".gif" => "image/gif",
            ".bmp" => "image/bmp",
            ".webp" => "image/webp",
            _ => "application/octet-stream" // Default binary
        };
    }

    public static bool IsValidImageFile(IFormFile file)
    {
        if (file == null || file.Length == 0)
            return false;

        string contentType = GetContentType(file);

        string[] allowedImageTypes =
        {
            "image/jpeg",
            "image/png",
            "image/gif",
            "image/bmp",
            "image/webp"
        };

        return allowedImageTypes.Contains(contentType);
    }

    public static bool HasValidFileName(string fileName)
    {
        if (string.IsNullOrWhiteSpace(fileName))
            return false;

        // Check for valid file name (avoid path traversal and invalid characters)
        var regex = new Regex(@"^[a-zA-Z0-9_\-\. ]+$");
        return regex.IsMatch(fileName);
    }

    public static string SanitizeFileName(string fileName)
    {
        // Remove any invalid characters
        var invalidChars = Path.GetInvalidFileNameChars();
        var sanitizedName = new string(fileName.Where(c => !invalidChars.Contains(c)).ToArray());

        // Additional sanitization
        return Regex.Replace(sanitizedName, @"[^\w\-\. ]", "_");
    }
}

File: go-api/src/helpers/file_helper.go
/*
* Copyright (C) Rodrigo Ferreira, All Rights Reserved
* Unauthorized copying of this file, via any medium is strictly prohibited
* Proprietary and confidential
 */

package helpers

import (
	"mime/multipart"
	"path/filepath"
	"regexp"
	"strings"
)

// GetContentType Determine the content type of a file
func GetContentType(file *multipart.FileHeader) string {
	// Try getting content type from the file header
	if file.Header != nil && len(file.Header["Content-Type"]) > 0 {
		return strings.ToLower(file.Header["Content-Type"][0])
	}

	// If the content type is not available, determine by extension
	extension := strings.ToLower(filepath.Ext(file.Filename))

	switch extension {
	case ".jpg", ".jpeg":
		return "image/jpeg"
	case ".png":
		return "image/png"
	case ".gif":
		return "image/gif"
	case ".bmp":
		return "image/bmp"
	case ".webp":
		return "image/webp"
	default:
		return "application/octet-stream" // Binário padrão
	}
}

// IsValidImageFile Verify if the file is a valid image file
func IsValidImageFile(file *multipart.FileHeader) bool {
	if file == nil || file.Size == 0 {
		return false
	}

	contentType := GetContentType(file)

	allowedImageTypes := []string{
		"image/jpeg",
		"image/png",
		"image/gif",
		"image/bmp",
		"image/webp",
	}

	for _, allowedType := range allowedImageTypes {
		if contentType == allowedType {
			return true
		}
	}

	return false
}

// HasValidFileName verifica se o nome do arquivo é válido
func HasValidFileName(fileName string) bool {
	if fileName == "" {
		return false
	}

	// Verificar nome de arquivo válido (evitar path traversal e caracteres inválidos)
	regex := regexp.MustCompile(`^[a-zA-Z0-9_\-\. ]+$`)
	return regex.MatchString(fileName)
}

// SanitizeFileName sanitiza o nome do arquivo
func SanitizeFileName(fileName string) string {
	// Remover caracteres inválidos
	regex := regexp.MustCompile(`[^\w\-\. ]`)
	return regex.ReplaceAllString(fileName, "_")
}
